libgadget/lenstools.c

```diff
@@ -0,0 +1,412 @@
+#include "lenstools.h"
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <fftw3.h> 
+
+#ifdef USE_CFITSIO
+#include "fitsio.h"
+#endif
+
+#include <string.h>
+#include "partmanager.h"
+#include "cosmology.h"
+#include "physconst.h"
+
+void linspace(double start, double stop, int num, double *result) {
+    double step = (stop - start) / (num - 1);
+    for (int i = 0; i < num; i++) {
+        result[i] = start + i * step;
+    }
+}
+
+double ***allocate_3d_array(int nx, int ny, int nz) {
+    // Allocate memory for the 3D array
+    double ***array = malloc(nx * sizeof(double **));
+    if (array == NULL) {
+        fprintf(stderr, "Memory allocation failed\n");
+        exit(EXIT_FAILURE);
+    }
```

Comment:
If you are using mymalloc the code auto-exits when malloc fails and you don't need the checks.